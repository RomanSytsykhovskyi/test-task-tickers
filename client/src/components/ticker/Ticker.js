import React, { useRef, useEffect, useMemo, useState } from "react";
import ExtraInfo from "./extraInfo/ExtraInfo";

//css
import "./style.css";

//imgs
//import crossOut from "./img/cross-out.png";

const noop = () => {};

const Ticker = ({ ticker = {}, hideHandler = noop, deleteHandler = noop }) => {
  const currentPrice = useMemo(() => Number(ticker.price), [ticker.price]);
  const prevPrice = useRef(0);

  const delta = useMemo(
    () => Number(currentPrice) - Number(prevPrice.current),
    [currentPrice]
  );
  const greenOrRed = delta >= 0 ? " green" : " red";

  const percent = useMemo(() => (100 * delta) / prevPrice.current, [delta]);

  const [isShown, setShown] = useState(false);

  useEffect(() => {
    prevPrice.current = currentPrice;
  });

  if (Object.is(ticker, {})) return null;

  return (
    <React.Fragment>
      <ul className="ticker">
        <li>
          <ul className="finance">
            <li>{ticker.ticker}: </li>
            <li className="price">Price: {currentPrice}</li>
            <li className={`delta${greenOrRed}`}>Change: {delta.toFixed(2)}</li>
            <li className={`percent${greenOrRed}`}>
              Percent: {`${percent.toFixed(2)} %`}
            </li>
            <li className="visibity">
              <button
                onClick={() => {
                  setShown(false);
                  hideHandler(ticker.ticker);
                }}
              >
                hide
              </button>
              <button
                onClick={() => {
                  setShown(false);
                  deleteHandler(ticker.ticker);
                }}
              >
                &#10008;
              </button>
            </li>
          </ul>
        </li>
        <ExtraInfo
          props={ticker}
          _yield={ticker.yield}
          isShown={isShown}
          setShown={setShown}
        />
      </ul>
    </React.Fragment>
  );
};

export default Ticker;
