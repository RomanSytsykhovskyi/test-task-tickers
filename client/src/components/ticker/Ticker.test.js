import React from "react";
import { render, screen } from "@testing-library/react";
import { unmountComponentAtNode } from "react-dom";
import useEvent from "@testing-library/user-event";

import Ticker from "./Ticker";

const ticker = {
  ticker: "AAPL",
  exchange: "NASDAQ",
  price: "246.55",
  change: "51.28",
  change_percent: "0.48",
  dividend: "0.32",
  yield: "1.34",
  last_trade_time: "2022-05-23T06:28:36.000Z",
};

let container = null;
beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe("It's a Ticker component", () => {
  const f = jest.fn();
  const deleteF = jest.fn();
  it("Ticker component renders", () => {
    render(<Ticker ticker={ticker} clickHandler={f} deleteHandler={deleteF} />);

    expect(screen.getByText(/AAPL/)).toBeInTheDocument();
  });

  it("Ticker component renders without props", () => {
    render(<Ticker />);

    expect(screen.queryByText(/AAPL/)).toBeNull();
  });

  it("Correct difference and percent", () => {
    const { container, rerender } = render(
      <Ticker ticker={ticker} clickHandler={f} deleteHandler={deleteF} />
    );

    const prevPrice = 246.55;
    expect(screen.getByText(`Price: ${prevPrice}`)).toBeInTheDocument();

    const currentPrice = 250;
    rerender(<Ticker ticker={{ ...ticker, price: `${currentPrice}` }} />);

    const delta = currentPrice - prevPrice;
    expect(container.getElementsByClassName("delta")[0].innerHTML).toBe(
      `Change: ${delta.toFixed(2)}`
    );

    expect(container.getElementsByClassName("percent")[0].innerHTML).toBe(
      `Percent: ${((100 * delta) / prevPrice).toFixed(2)} %`
    );
  });

  it("hideHandler works correctly", () => {
    render(<Ticker ticker={ticker} hideHandler={f} deleteHandler={deleteF} />);

    useEvent.click(screen.getAllByRole("button")[0]);

    expect(f).toHaveBeenCalledTimes(1);
  });

  it("deleteHandler works correctly", () => {
    render(<Ticker ticker={ticker} clickHandler={f} deleteHandler={deleteF} />);

    useEvent.click(screen.getAllByRole("button")[1]);

    expect(deleteF).toHaveBeenCalledTimes(1);
  });

  it("Style for increasing price works", () => {
    const { container, rerender } = render(
      <Ticker ticker={ticker} clickHandler={f} deleteHandler={deleteF} />
    );

    const currentPrice = 250;
    rerender(<Ticker ticker={{ ...ticker, price: `${currentPrice}` }} />);

    expect(container.getElementsByClassName("delta")[0]).toHaveClass("green");
  });

  it("Style for decreasing price works", () => {
    const { container, rerender } = render(
      <Ticker ticker={ticker} clickHandler={f} deleteHandler={deleteF} />
    );

    const currentPrice = 230;
    rerender(<Ticker ticker={{ ...ticker, price: `${currentPrice}` }} />);

    expect(container.getElementsByClassName("delta")[0]).toHaveClass("red");
  });

  it("Ticker snapshot", () => {
    const renderedTicker = render(
      <Ticker ticker={ticker} clickHandler={f} deleteHandler={deleteF} />
    );

    expect(renderedTicker).toMatchSnapshot();
  });
});
