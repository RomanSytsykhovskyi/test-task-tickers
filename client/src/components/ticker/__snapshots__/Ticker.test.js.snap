// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It's a Ticker component Ticker snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div />
    <div>
      <ul
        class="ticker"
      >
        <li>
          <ul
            class="finance"
          >
            <li>
              AAPL
              : 
            </li>
            <li
              class="price"
            >
              Price: 
              246.55
            </li>
            <li
              class="delta green"
            >
              Change: 
              246.55
            </li>
            <li
              class="percent green"
            >
              Percent: 
              Infinity %
            </li>
            <li
              class="visibity"
            >
              <button>
                hide
              </button>
              <button>
                ✘
              </button>
            </li>
          </ul>
        </li>
        <li
          class="isShown"
        >
          <button>
            ✚
          </button>
        </li>
        <li
          class="info hidden"
        >
          <ul>
            <li>
              Exchange:
              <br />
              NASDAQ
            </li>
            <li>
              Change:
              <br />
              51.28
            </li>
            <li>
              Change percent:
              <br />
              0.48
            </li>
            <li>
              Divedend:
              <br />
              0.32
            </li>
            <li>
              Yield:
              <br />
              1.34
            </li>
            <li>
              Last trade time:
              <br />
              2022-05-23T06:28:36.000Z
            </li>
          </ul>
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul
      class="ticker"
    >
      <li>
        <ul
          class="finance"
        >
          <li>
            AAPL
            : 
          </li>
          <li
            class="price"
          >
            Price: 
            246.55
          </li>
          <li
            class="delta green"
          >
            Change: 
            246.55
          </li>
          <li
            class="percent green"
          >
            Percent: 
            Infinity %
          </li>
          <li
            class="visibity"
          >
            <button>
              hide
            </button>
            <button>
              ✘
            </button>
          </li>
        </ul>
      </li>
      <li
        class="isShown"
      >
        <button>
          ✚
        </button>
      </li>
      <li
        class="info hidden"
      >
        <ul>
          <li>
            Exchange:
            <br />
            NASDAQ
          </li>
          <li>
            Change:
            <br />
            51.28
          </li>
          <li>
            Change percent:
            <br />
            0.48
          </li>
          <li>
            Divedend:
            <br />
            0.32
          </li>
          <li>
            Yield:
            <br />
            1.34
          </li>
          <li>
            Last trade time:
            <br />
            2022-05-23T06:28:36.000Z
          </li>
        </ul>
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
